### ISSUE TITLE
Suggestion: enhance Method Load event records to indicate what kind of code the jit produced

### ISSUE THREAD

**Issue by AndyAyersMS** [dotnet/coreclr#14458]
There is already some diversity in jit codegen. For instance the jit can be instructed, or may on its own decide whether to generate optimized or minimally optimized code, or use AVX2/SSE2 and certain HW instructions, or to enable vector codegen, etc.

This diversity is about to increase substantially, with the advent of fine-grained HW intrinsics and tiering.

For diagnostic purposes it would be useful to know key attributes of the code the jit generated for each method, for instance:
* which target-specific triggers (that is, which IsSupported intrinsics) were encountered during codegen, and how they evaluated (could be tri state: yes/no/maybe, depending on where we end up with evaluation of these when prejitting)
* what codegen mode the jit used for the method (minops / debug / tier0 / fullopt)
* what codegen mode was requested by the VM
* some kind of HW checksum to fingerprint the generated native code. Maybe via some abstracted ISA representation to avoid being overly sensitive on absolute or relative address literals, though there is also some merit in being able to use the checksum to re-validate existing code.
* whether some out-of-band signal (eg complus setting) influenced jit behavior
* whether the method was optimized with IBC, and if so, whether it was classified as hot/cold
* if prejitted, whether prejitting was fragile or used R2R codegen (this is perhaps available by looking back at the module load event?)

We have a subset of this now, but it is not comprehensive.

Having this would make it easier to verify that the jit generates the intended kinds of code.

The fingerprinting could come in handy for quick automated "diff" testing that would work in release or as part of an automated diagnostics package for root-causing correctness or performance regressions. We can do some of this now with the method size information but a hash would be more collision-proof.

cc @noahfalk

**Comment by noahfalk** [comment]
This all sounds like goodness to me, but I agree we need to be careful about the prejitted case. In its ideal form we wouldn't run any preparation code at all on a per-method basis for prejitted code. Having tooling infer the information from a module scoped event seems pretty workable though.

I think the next question would be which places in our code know all of this information. I'm guessing some of it is only known by the JIT dll, in which case we need to figure out what the pathway is to get that info from the JIT to event pipeline. My intuitive notion would be to add a call to the EE-JIT interface so that the JIT remains agnostic of what kind of event pipeline is being used. Then the runtime can forward it to various event streams that are configured to consume it (ETW, in-proc profiler, ICorProfiler, logging, etc).

I'm guessing in the short term at least the intended consumer of the data would be our team using PerfView or other custom TraceEvent derived tools?

**Comment by AndyAyersMS** [comment]
Yes, in the short term this would mainly be used to validate that the jit is creating the code we expect, and facilitate and speed up our ability to detect "diffs" in jit codegen relatively quickly.

For the latter, note our current "diff" technology relies heavily on either prejitting (as in the jit-diff tooling) or on capture and replay of jit-vm interface traffic (as in SPMI).

The former currently requires a fairly slow Checked jit and is only reliably effective for prejit codegen (and so excludes dynamic, vector and HW dependent codegen, and (in R2R) misses out on jitting of many generic instantations).

The latter produces fragile and relatively large files to hold all the interface chatter.

An appealing alternative is to run a test suite in A/B mode (say old jit/new jit) with release bits with the CLR event stream enabled and postprocess this down to a "jit behavior" trace which can be diffed. This could be run as part of CI and summarized so jit devs (or fx devs) could get some idea of the codegen impact of a change. If/when diffs are found one could circle back and rerun the diffing parts with Checked jits to expose the details.

We would also use this to do Release/Check consistency validation in CI, which would be very nice to have.

Some of this we can at now using the method size field, so it might be worth building up the tooling and some prototype workflows and see how well these ideas hold up in practice.
